Pluggable Symbol finder API, with Python support

brenns10 opened the pull request:
This implements a pluggable Symbol Finder API, to complement the Object and Type finding APIs. There is also a Python binding.

I think it's a bit easier to review this by reading each commit rather than the full diff, but I don't know how easy Github makes that.

I'll highlight a few architectural points here that may merit discussion or debate.

1. The symbol finder API has a "result builder", because the API allows requesting exactly one, or multiple symbols. I think this is pretty important and wanted to avoid the additional allocation for an array of size one. The result builder helps smooth over the API and simplifies error handling.
2. Symbols now contain ownership info to determine how Drgn should treat the string (should it be copied? freed? or left alone). This wasn't necessary when passing symbols from drgn->python, because the names had the lifetime of the Program (`LIFETIME_STATIC` in this new system). Strings passed from python->drgn are subject to garbage collection, so they need to be copied and drgn will manage their lifetime.
3. ~The API exposes `drgn_module` as a struct forward declaration, it's unused at this time, except by the ELF symbol finder.~

There are added tests which do demonstrate use from Python. You can also see the current [kallsyms_finder](https://github.com/brenns10/drgn/commits/kallsyms_finder) branch for an example of using the C API.

(PR description is re-edited now that I've moved this out of draft state.)

brenns10 commented:
@osandov I think the Symbol finder API is now ready to review! I figured I'd continue using this old PR for the branch, but I did update the description so it's all relevant.

osandov commented on libdrgn/python/symbol.c:
```diff
@@ -3,23 +3,83 @@
+	}
+
+	sym->name_owned = true;
+	sym->name = strdup(name);
```
(This is a higher-level comment that may obsolete some of the other comments I left here.)

Instead of `strdup`ing the name, I think we can hold a reference to the original name `PyObject`. The reference we hold on `Symbol::prog` is only used to keep the name alive; we could replace that with a more general `PyObject *` that we need to hold for the lifetime of the symbol. Something along these lines:

```diff
diff --git a/libdrgn/python/drgnpy.h b/libdrgn/python/drgnpy.h
index 6f5fc6d5..65284838 100644
--- a/libdrgn/python/drgnpy.h
+++ b/libdrgn/python/drgnpy.h
@@ -173,7 +173,8 @@ typedef struct {
 
 typedef struct {
 	PyObject_HEAD
-	Program *prog;
+	// Reference we need to hold to keep the symbol name alive.
+	PyObject *name_obj;
 	struct drgn_symbol *sym;
 } Symbol;
 
@@ -288,7 +289,7 @@ Program *program_from_core_dump(PyObject *self, PyObject *args, PyObject *kwds);
 Program *program_from_kernel(PyObject *self);
 Program *program_from_pid(PyObject *self, PyObject *args, PyObject *kwds);
 
-PyObject *Symbol_wrap(struct drgn_symbol *sym, Program *prog);
+PyObject *Symbol_wrap(struct drgn_symbol *sym, PyObject *prog);
 
 PyObject *Thread_wrap(struct drgn_thread *drgn_thread);
 
diff --git a/libdrgn/python/program.c b/libdrgn/python/program.c
index c2103f48..a43ec63a 100644
--- a/libdrgn/python/program.c
+++ b/libdrgn/python/program.c
@@ -1012,7 +1012,7 @@ static PyObject *Program_symbols(Program *self, PyObject *args)
 		return NULL;
 	}
 	for (size_t i = 0; i < count; i++) {
-		PyObject *pysym = Symbol_wrap(symbols[i], self);
+		PyObject *pysym = Symbol_wrap(symbols[i], (PyObject *)self);
 		if (!pysym) {
 			/* Free symbols which aren't yet added to list. */
 			drgn_symbols_destroy(symbols, count);
@@ -1048,7 +1048,7 @@ static PyObject *Program_symbol(Program *self, PyObject *arg)
 	}
 	if (err)
 		return set_drgn_error(err);
-	ret = Symbol_wrap(sym, self);
+	ret = Symbol_wrap(sym, (PyObject *)self);
 	if (!ret) {
 		drgn_symbol_destroy(sym);
 		return NULL;
diff --git a/libdrgn/python/stack_trace.c b/libdrgn/python/stack_trace.c
index d738da14..125cc984 100644
--- a/libdrgn/python/stack_trace.c
+++ b/libdrgn/python/stack_trace.c
@@ -218,7 +218,7 @@ static PyObject *StackFrame_symbol(StackFrame *self)
 	err = drgn_stack_frame_symbol(self->trace->trace, self->i, &sym);
 	if (err)
 		return set_drgn_error(err);
-	PyObject *ret = Symbol_wrap(sym, prog);
+	PyObject *ret = Symbol_wrap(sym, (PyObject *)prog);
 	if (!ret) {
 		drgn_symbol_destroy(sym);
 		return NULL;
diff --git a/libdrgn/python/symbol.c b/libdrgn/python/symbol.c
index 0ea395e2..a625c295 100644
--- a/libdrgn/python/symbol.c
+++ b/libdrgn/python/symbol.c
@@ -7,26 +7,13 @@
 #include "drgnpy.h"
 #include "symbol.h"
 
-PyObject *Symbol_wrap(struct drgn_symbol *sym, Program *prog)
+PyObject *Symbol_wrap(struct drgn_symbol *sym, PyObject *name_obj)
 {
-	Symbol *ret;
-
-	if (!sym->name_owned && !prog) {
-		/* If a symbol does not own its name, then it implicitly
-		 * shares the lifetime of its associate program. So, we
-		 * must have it here. Symbols created via Python (Symbol_new())
-		 * always own their names. */
-		PyErr_SetString(PyExc_ValueError,
-				"Creating symbol with unowned name,"
-				" without program reference");
-		return NULL;
-	}
-	ret = call_tp_alloc(Symbol);
+	Symbol *ret = call_tp_alloc(Symbol);
 	if (ret) {
 		ret->sym = sym;
-		ret->prog = prog;
-		if (prog)
-			Py_INCREF(prog);
+		ret->name_obj = name_obj;
+		Py_INCREF(name_obj);
 	}
 	return (PyObject *)ret;
 }
@@ -35,7 +22,7 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 {
 	struct drgn_symbol *sym;
 	static char *keywords[] = {"name", "address", "size", "binding", "kind", NULL};
-	const char *name;
+	PyObject *name_obj;
 	struct index_arg address = {}, size = {};
 	struct enum_arg binding = {
 		.type = SymbolBinding_class,
@@ -48,8 +35,8 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 	if (!sym)
 		return set_drgn_error(&drgn_enomem);
 
-	if (!PyArg_ParseTupleAndKeywords(args, kwds, "sO&O&O&O&:Symbol", keywords,
-					 &name,
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!O&O&O&O&:Symbol",
+					 keywords, &PyUnicode_Type, &name_obj,
 					 index_converter, &address,
 					 index_converter, &size,
 					 enum_converter, &binding,
@@ -57,18 +44,17 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 		free(sym);
 		return NULL;
 	}
+	const char *name = PyUnicode_AsUTF8(name_obj);
+	if (!name)
+		return NULL;
 
-	sym->name_owned = true;
-	sym->name = strdup(name);
-	if (!sym->name) {
-		free(sym);
-		return set_drgn_error(&drgn_enomem);
-	}
+	sym->name_owned = false;
+	sym->name = name;
 	sym->address = address.uvalue;
 	sym->size = size.uvalue;
 	sym->binding = binding.value;
 	sym->kind = kind.value;
-	PyObject *ret = Symbol_wrap(sym, NULL);
+	PyObject *ret = Symbol_wrap(sym, name_obj);
 	if (!ret) {
 		drgn_symbol_destroy(sym);
 	}
@@ -78,8 +64,7 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 static void Symbol_dealloc(Symbol *self)
 {
 	drgn_symbol_destroy(self->sym);
-	if (self->prog)
-		Py_DECREF(self->prog);
+	Py_XDECREF(self->name_obj);
 	Py_TYPE(self)->tp_free((PyObject *)self);
 }
 
```

This name is not owned by the `drgn_symbol` or the `drgn_program`, so either we need a third state or `drgn_symbol_copy` needs to always copy the name.

brenns10 replied:
This is a great suggestion, thanks!

Sadly the string situation is all sorts of messed up still, even with this improvement. For symbols which are created internally to Drgn via C code, we can usually guarantee the lifetime of the string to be the same as the lifetime as the program, so the string copying situation looks like this:

1. Usually no copy on creation, due to the lifetime of the underlying symbol data source.
2. When you provide the symbol via the symbol finder API, no copy. (one malloc for the `drgn_symbol`)
3. When the corresponding Python Symbol object is created, it takes ownership of the `drgn_symbol` so no copy happens here, for the name or the struct.
4. When Python code accesses the name, we use `PyUnicode_FromString` which must make a copy, but that's truly unavoidable.

But if you're a Python Symbol finder:

1. One strdup on construction of the Symbol object, and we also malloc a new `drgn_symbol`. (Your improvement eliminates the strdup)
2. We use `drgn_symbol_copy()` and we `malloc()` a new `drgn_symbol` as well. We can't guarantee the lifetime of this name, so we'd need a strdup for the name. And you're right, we'd need a new state for the symbol to represent "don't free the name, but duplicate it when you copy it".
3. If the symbol is getting copied back out to a Python client, then `Symbol_wrap()` will end up taking ownership of the `drgn_symbol` and so no copy happens there. But the Python client would then copy the string on access to the name field, as above.

So for the first case, one copy of the string element itself, one malloc of the drgn_symbol. In the second case, the current branch would result in the name field being copied 3 times in order to be used, and the drgn_symbol structure is allocated twice. With your improvement, the 3 name copies becomes 2, and we still have allocated the `drgn_symbol` twice. It's not great :(

I have some wild idea where we do something like this instead of your approach:
```
diff --git a/libdrgn/python/program.c b/libdrgn/python/program.c
index c2103f4..091251b 100644
--- a/libdrgn/python/program.c
+++ b/libdrgn/python/program.c
@@ -547,16 +547,21 @@ static struct drgn_error *py_symbol_find_fn(const char *name, uint64_t addr,
 			err = drgn_error_from_python();
 			goto out;
 		}
-		_cleanup_free_ struct drgn_symbol *sym = malloc(sizeof(*sym));
-		if (!sym) {
-			err = &drgn_enomem;
-			goto out;
-		}
-		err = drgn_symbol_copy(sym, ((Symbol *)item)->sym);
-		if (err)
-			goto out;
+		if (Py_REFCNT(item) == 1) {
+			err = drgn_symbol_result_builder_add(builder, ((Symbol *)item)->sym);
+			((Symbol *)item)->sym = NULL;
+		} else {
+			_cleanup_free_ struct drgn_symbol *sym = malloc(sizeof(*sym));
+			if (!sym) {
+				err = &drgn_enomem;
+				goto out;
+			}
+			err = drgn_symbol_copy(sym, ((Symbol *)item)->sym);
+			if (err)
+				goto out;
 
-		err = drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym));
+			err = drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym));
+		}
 	}
 
 out:
```

We have to copy the string on creation of the Python Symbol object, and allocate a `drgn_symbol`. But if the Python symbol finder function simply returns the symbol without retaining a reference, we can take ownership of it. We can avoid allocating a second `drgn_symbol` and avoid that copy. The result is that the name is copied twice, and `drgn_symbol` allocated once.

I don't know if this is better though. Just an alternative which avoids the third state for the `drgn_symbol` object, and reduces the amount we allocate a `drgn_symbol`.

osandov replied:
Your idea is clever, although I think the `Py_REFCNT(item) == 1` check isn't quite enough. Consider:

```python3
my_symbols = [Symbol(...)]
def my_symbol_finder(name, address, first):
    return my_symbols
prog.add_symbol_finder(my_symbol_finder)
prog.symbol(...)
print(my_symbols[0])
```

I.e., there's only one reference on the symbol, but it can still be accessed via the list. So you probably also need a `Py_REFCNT(list) == 1` check. At this point it feels like we're playing with fire a bit :)

I don't love my third state idea either, though, so I'll leave it up to you which way to go.

osandov reviewed (changes requested):
Sorry for the delay, this is looking pretty good. The biggest architectural comment I have is regarding the `strdup` in `Symbol_wrap`; the rest are more about the implementation or style.

brenns10 commented on libdrgn/python/program.c:
```diff
@@ -481,6 +481,89 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
+		goto out;
+	}
+
+	for (size_t i = 0; i < len; i++) {
+		PyObject *item = PyList_GET_ITEM(obj, i);
+		if (!PyObject_TypeCheck(item, &Symbol_type)) {
+			PyErr_SetString(PyExc_ValueError,
+					"symbol find callback elements must be type Symbol");
+			err = drgn_error_from_python();
+			goto out;
+		}
+		_cleanup_free_ struct drgn_symbol *sym = malloc(sizeof(*sym));
+		if (!sym) {
+			err = &drgn_enomem;
+			goto out;
+		}
+		err = drgn_symbol_copy(sym, ((Symbol *)item)->sym);
+		if (err)
+			goto out;
+
+		err = drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym));
+	}
+
+out:
+	PyGILState_Release(gstate);
+	return err;
+}
```
This is embarassing, there's a really obvious and really bad memory leak here. We own the reference to the return value of the Python symbol finder function. I need to decref `obj` here in order to free all the symbols and the containing list -- regardless of the success or failure of the function.

brenns10 replied:
Oh wait, wow, never mind, this is just illustrating the exact point you made above about putting the `_cleanup_pydecref_` declaration closest to where the object is used...

osandov commented on _drgn.pyi:
```diff
@@ -448,7 +450,39 @@ class Program:
+            use the built-in ELF symbol finder first. If this fails, it will
+            continue to call each callback in the order described.
+
+        :param fn: Callable taking address, name, and flag, and returning a list
```
The order of address vs name is inconsistent between here and the type annotation of `fn`. Can we also name `flag` something else, like `first` or the converse, `all`?

brenns10 replied:
It's kinda sad that Python's type annotations don't let you give parameter names like a C typedef would. I updated the comment here to read "callable taking name, address, and first flag, ...". I also updated some other references to call it the "first flag" as well.

osandov commented on _drgn.pyi:
```diff
@@ -448,7 +450,39 @@ class Program:
+        """
+        ...
+    def add_symbol_finder(
+        self, fn: Callable[[Optional[str], Optional[int], bool], List[Symbol]]
```
We should allow this to return any `Sequence[Symbol]` instead of specifically a `List[Symbol]`. See [`PySequence_Fast`](https://docs.python.org/3/c-api/sequence.html#c.PySequence_Fast) for one way to implement that. We use that in a few places in the Python bindings already.

osandov commented:
Fyi, I merged the first commit in this PR ("Correct types/docs for object+type finder API").

brenns10 commented on libdrgn/drgn.h.in:
```diff
@@ -904,9 +915,93 @@ struct drgn_error *drgn_program_find_symbols_by_name(struct drgn_program *prog,
+drgn_program_add_symbol_finder(struct drgn_program *prog,
+			       drgn_find_symbol_fn fn, void *arg);
+
+/**
+ * Unregister a symbol finding callback
+ *
+ * If a symbol finder is found with the given @a fn and @a arg, removes it.
+ * Otherwise, does nothing.
+ *
+ * @param[in] fn Symbol search function
+ * @param[in] arg Callback argument
+ */
+void drgn_program_remove_symbol_finder(struct drgn_program *prog,
+				       drgn_find_symbol_fn fn, void *arg);
```
No, not really, I'll follow the same pattern as the object and type finders.

brenns10 commented on libdrgn/python/symbol.c:
```diff
@@ -3,23 +3,83 @@
+		return NULL;
+	}
+
+	sym->name_owned = true;
+	sym->name = strdup(name);
+	if (!sym->name) {
+		free(sym);
+		return set_drgn_error(&drgn_enomem);
+	}
+	sym->address = address.uvalue;
+	sym->size = size.uvalue;
+	sym->binding = binding.value;
+	sym->kind = kind.value;
```
That's a good idea, I hadn't really thought about public / private APIs and how this API would get consumed by external C clients. I'll add it.

brenns10 commented on libdrgn/program.c:
```diff
@@ -1762,57 +1773,69 @@ struct drgn_error *drgn_error_symbol_not_found(uint64_t address)
+		 * symbol. Otherwise, save the symbol only if we haven't
+		 * found another symbol.
+		 */
+		if (sym->binding == DRGN_SYMBOL_BINDING_GLOBAL ||
+			sym->binding == DRGN_SYMBOL_BINDING_UNIQUE ||
+			sym->binding == DRGN_SYMBOL_BINDING_WEAK ||
+			!drgn_symbol_result_builder_count(arg->builder)) {
```
Yep, I'll reorganize that a bit

brenns10 commented on libdrgn/program.c:
```diff
@@ -1825,164 +1848,205 @@ static int symbols_search_cb(Dwfl_Module *dwfl_module, void **userdatap,
+drgn_program_symbols_search(struct drgn_program *prog, const char *name,
+			    uint64_t addr, struct drgn_module *module,
+			    enum drgn_find_symbol_flags flags,
+			    struct drgn_symbol_result_builder *ret)
```
I think the reason I had it written as an out parameter so that I could include the repetitive error handling logic in this function. And it's a "result builder" rather than an array and length, because I wanted to avoid the extra allocation of a small array for the case where you're looking up a single symbol.

I see why it's a bit unidiomatic though. I'll see if there's a nicer way to do it.

brenns10 commented on _drgn.pyi:
```diff
@@ -448,7 +450,39 @@ class Program:
+        finder returns a result. Otherwise, all callbacks will be called, and
+        all results will be returned.
+
+        .. note::
+
+            There is one case where callback order is not respected: drgn's
+            internal symbol resolution. If drgn has debuginfo already loaded for
+            a particular memory address, it will take a shortcut and directly
+            use the built-in ELF symbol finder first. If this fails, it will
+            continue to call each callback in the order described.
+
```
This note is outdated. I removed it for the second revision, but it shouldn't have even been included in the first revision. If you look at `drgn_find_symbol_by_address_internal()`, it also uses the symbol finder callbacks.

I wrote this note before I had incorporated a `struct drgn_module *` into the symbol finder API. Without that pointer, it was impossible to implement the shortcut for address lookup with `dwfl_addrmodule()`. So I ended up leaving `drgn_find_symbol_by_address_internal()` to use the internal ELF symbol code. But thankfully that's all gone now: the only difference between the internal and external one as that the internal one does not return a lookup error.

brenns10 commented on libdrgn/python/program.c:
```diff
@@ -481,6 +481,89 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
+
+	size_t len = PyList_GET_SIZE(obj);
+	if (len > 1 && (flags & DRGN_FIND_SYM_ONE))  {
+		PyErr_SetString(PyExc_ValueError,
+				"symbol find callback returned multiple elements, but one was requested");
+		err = drgn_error_from_python();
```
I don't know what exactly I thought would happen as a result of this odd pattern, but what actually happens is:
```
Traceback (most recent call last):
  File "/home/stepbren/repos/drgn/tests/test_symbol.py", line 319, in test_args_single_with_many_results
    self.prog.symbol(0xFFFF2008)
Exception: ValueError: symbol find callback returned multiple elements, but one was requested
```
I'm going to directly return a drgn error here, and add a test covering this case.

brenns10 commented:
Ok, I went ahead and rebased this branch on top of the latest main. That did involve some changes regarding the `prog->dbinfo` change you mentioned, and now the symbol finder API looks more or less exactly like the type/object finders.

I also tried to do a better review of my commits prior to sending them, and caught a few issues. I marked them so you could see what they were.

osandov reviewed (commented):
I'm finally getting back to this. I only re-reviewed patch 1 so far, but I just wanted to submit these before I head out for the day. Feel free to wait for me to finish looking at the rest. Thanks for your continued work on this!

osandov commented on libdrgn/program.h:
```diff
@@ -353,18 +353,18 @@ struct drgn_error *drgn_program_cache_prstatus_entry(struct drgn_program *prog,
- * Like @ref drgn_program_find_symbol_by_address(), but @p ret is already
- * allocated, we may already know the module, and doesn't return a @ref
- * drgn_error.
+ * Like @ref drgn_program_find_symbol_by_address(), but does not return a lookup
+ * error if the symbol was not found.
```
It'd be nice to clarify that it returns `NULL` in `ret` instead if the symbol is not found.

osandov commented on libdrgn/program.h:
```diff
@@ -353,18 +353,18 @@ struct drgn_error *drgn_program_cache_prstatus_entry(struct drgn_program *prog,
  *
  * @param[in] module Module containing the address. May be @c NULL, in which
  * case this will look it up.
  * @return Whether the symbol was found.
```
Needs an update.

osandov commented on libdrgn/stack_trace.c:
```diff
@@ -109,6 +109,7 @@ drgn_stack_trace_num_frames(struct drgn_stack_trace *trace)
 LIBDRGN_PUBLIC struct drgn_error *
 drgn_format_stack_trace(struct drgn_stack_trace *trace, char **ret)
 {
+	struct drgn_error *err = NULL;
```
No need to initialize this. Let's allow the compiler to warn us if we forget to set this.
```suggestion
	struct drgn_error *err;
```

osandov commented on libdrgn/stack_trace.c:
```diff
@@ -173,6 +174,7 @@ drgn_format_stack_frame(struct drgn_stack_trace *trace, size_t frame, char **ret
 {
 	STRING_BUILDER(str);
 	struct drgn_register_state *regs = trace->frames[frame].regs;
+	struct drgn_error *err = NULL;
```
Ditto.
```suggestion
	struct drgn_error *err;
```

osandov commented on libdrgn/symbol.h:
```diff
@@ -34,6 +35,12 @@ struct drgn_symbol_result_builder {
 };
 
+#define _cleanup_symbol_ _cleanup_(drgn_symbol_cleanup)
+static inline void drgn_symbol_cleanup(void *p)
+{
+	drgn_symbol_destroy(*(struct drgn_symbol **)p);
```
This can be type-safe:
```suggestion
static inline void drgn_symbol_cleanup(struct drgn_symbol **p)
{
	drgn_symbol_destroy(*p);
```
```

osandov commented on libdrgn/symbol.h:
```diff
@@ -16,6 +16,7 @@ struct drgn_symbol {
 	uint64_t size;
 	enum drgn_symbol_binding binding;
 	enum drgn_symbol_kind kind;
+	enum drgn_lifetime name_lifetime;
```
We should add `__attribute__((__packed__))` to the definitions of all of these `enum`s to avoid growing this structure. That can probably be its own preparatory commit for the existing `drgn_symbol_binding` and `drgn_symbol_kind`, then you can include it when you add the attribute in this commit.

osandov commented on libdrgn/drgn.h:
```diff
@@ -2983,6 +2983,44 @@ enum drgn_symbol_kind {
+enum drgn_lifetime {
+	/**
+	 * DRGN_LIFETIME_STATIC: the object is guaranteed to outlive the
+	 * drgn_program itself.  Drgn will not free or copy the object.
+	 */
+	DRGN_LIFETIME_STATIC,
+	/**
+	 * DRGN_LIFETIME_EXTERNAL: the object is externally managed. It will
+	 * live as long as the object it is associated with, but may be freed
+	 * after.  Drgn will never free the object. If drgn must copy a data
+	 * structure, the object will be duplicated, and drgn will own the new
+	 * object.
+	 */
+	DRGN_LIFETIME_EXTERNAL,
+	/**
+	 * DRGN_LIFETIME_OWNED: the object lifetime is managed by Drgn. It
```
`s/Drgn/drgn/g`

osandov commented on libdrgn/drgn.h:
```diff
@@ -2983,6 +2983,44 @@ enum drgn_symbol_kind {
 };
 
+/** Describes the lifetime of an object provided to drgn */
+enum drgn_lifetime {
```
:heart: :heart: :heart: This approach is way cleaner and more generic than I expected. Great job.

brenns10 replied:
Thanks!

osandov commented on libdrgn/python/program.c:
```diff
@@ -477,6 +477,75 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
+					    enum drgn_find_symbol_flags flags,
+					    void *data, struct drgn_symbol_result_builder *builder)
+{
+	struct drgn_error *err = NULL;
```
```suggestion
	struct drgn_error *err;
```

brenns10 replied:
In this case it would be possible to return an uninitialized `err` if the loop body never executed (because the Python code returned an empty list).

But I think the correct solution is to explicitly `return NULL` at the end of this function, and move the declaration of `err` into the body of the loop since it is only used there.

osandov commented on libdrgn/python/program.c:
```diff
@@ -477,6 +477,75 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
+		if (err)
+			return err;
+
+		err = drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym));
```
This leaks `sym` if `drgn_symbol_result_builder_add` fails.

brenns10 replied:
Yeah, this one is funny since we've already used `no_cleanup_ptr(sym)`, we need to free it. Kinda clunky.

brenns10 commented on libdrgn/symbol.c:
```diff
@@ -73,3 +73,61 @@ LIBDRGN_PUBLIC bool drgn_symbol_eq(struct drgn_symbol *a, struct drgn_symbol *b)
+
+DEFINE_VECTOR_FUNCTIONS(symbolp_vector);
+
+LIBDRGN_PUBLIC struct drgn_error *
```
I think it makes sense to switch to the convention. Especially since it would have made explicit the cases below where you caught some leaks in error cases: I would have needed to handle the error and probably realized I needed to free a symbol.

brenns10 replied:
Haha, this even revealed another leak in the kallsyms symbol finder.

I think I just assumed that on error, the builder would free the symbol. Which may be convenient in some cases, but isn't a good generic API.

brenns10 commented on libdrgn/program.c:
```diff
@@ -1798,57 +1799,76 @@ struct drgn_error *drgn_error_symbol_not_found(uint64_t address)
+		 * symbol. Otherwise, save the symbol only if we haven't
+		 * found another symbol.
+		 */
+		if (!(binding == STB_GLOBAL || binding == STB_GNU_UNIQUE ||
+		      binding == STB_WEAK ||
+		      !drgn_symbol_result_builder_count(arg->builder)))
```
yuuuuup this has gotten me several times on re-read. Love the rewrite, shipit!

brenns10 commented on libdrgn/symbol.c:
```diff
@@ -73,3 +73,61 @@ LIBDRGN_PUBLIC bool drgn_symbol_eq(struct drgn_symbol *a, struct drgn_symbol *b)
+void drgn_symbol_result_builder_array(struct drgn_symbol_result_builder *builder,
+				      struct drgn_symbol ***syms_ret, size_t *count_ret)
+{
+	symbolp_vector_steal(&builder->vector, syms_ret, count_ret);
```
You know, I think I assumed that `_steal()` would do that for me. Updated.

brenns10 commented on libdrgn/drgn.h:
```diff
@@ -927,9 +938,81 @@ struct drgn_error *drgn_program_find_symbols_by_name(struct drgn_program *prog,
+ */
+typedef struct drgn_error *
+(*drgn_find_symbol_fn)(const char *name, uint64_t addr,
+		       struct drgn_module *module,
```
From our discussion yesterday, we will be removing `struct drgn_module *` from the API. Marking this comment here so I don't forget to actually do it!

brenns10 commented on libdrgn/drgn.h:
```diff
@@ -501,6 +501,15 @@ struct drgn_thread;
  */
 struct drgn_program; // IWYU pragma: export
 
+/**
+ * @struct drgn_module
+ *
+ * Represents a module within the program. For the Linux kernel, this could be
+ * the main kernel (vmlinux) or a kernel module. For a userspace program, this
+ * could be the executable or a shared library.
+ */
+struct drgn_module; // IWYU pragma: export
+
```
And of course this will mean we get rid of the declaration in `drgn.h` too.

brenns10 commented on libdrgn/program.h:
```diff
@@ -364,6 +366,23 @@ bool drgn_program_find_symbol_by_address_internal(struct drgn_program *prog,
+ * Implementation of the Symbol finder API, based on ELF symbols
+ */
+struct drgn_error *
+elf_symbols_search(const char *name, uint64_t addr, struct drgn_module *module,
```
Yeah, I will add that as a follow-up commit, to keep the history clean.

brenns10 commented:
Alright, I believe I've addressed everything from your review on Github as well as our discussion. The `drgn_module` is now removed from the API as discussed, which I recall was the only major change we decided on in the discussion. That does reduce the extent of these changes nicely, since I don't need to touch all the callers of the symbol finding APIs.

Normally I like to keep my branches rebased on top of the latest `main`, but it seems easiest for me to leave the base of the branch where it is, so you can more easily compare the revisions:

https://github.com/osandov/drgn/compare/43cf71669fcbc5f66c38f5904326de53c3dd346e..711399a7a159d0e9a772d16562c814382a86fc2f

I did do a quick compile & test for each commit in the branch to make sure all of my changes were "fixed up" into the correct commits. There are also two new commits on the branch as well: one for the `__attribute__((__packed__))` change, and one to move the ELF symbol finders into `debug_info.c`.

osandov commented on _drgn.pyi:
```diff
@@ -453,6 +453,32 @@ class Program:
+        """
+        Register a callback for finding symbols in the program.
+
+        The callback should take three arguments: a search name, a search
+        address, and a boolean flag 'first' indicating whether to return only
+        the first match. When the 'first' flag is True, the callback should
+        return just one :class:`Symbol`. When the flag is False, the callback
+        should return a list of all matching :class:`Symbol`\\ s. Both the name
+        and address arguments are optional. If both are provided, then the
+        result(s) should match both. If neither are provided, the finder should
+        return all available symbols.
+
+        Callbacks are called in reverse order of the order they were added
+        (i.e,, the most recently added callback is called first). When the
+        'first' flag is set, the search will short-circuit after the first
+        finder which returns a result, and subsequent finders will not be
+        called. Otherwise, all callbacks will be called, and all results will be
+        returned.
+
+        :param fn: Callable taking name, address, and 'first' flag, and
+            returning a sequence of :class:`Symbol`\\ s.
```
Could you rephrase this to call the `first` argument `one` instead, both to match the libdrgn documentation and indicate that it's not necessarily the first match, but the "best" match?

osandov commented on libdrgn/python/program.c:
```diff
@@ -477,6 +477,73 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
+
+	size_t len = PySequence_Fast_GET_SIZE(obj);
+	if (len > 1 && (flags & DRGN_FIND_SYM_ONE))  {
+		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+					 "symbol find callback returned multiple elements, but one was requested");
```
Should this use `PyErr_SetString(PyExc_ValueError, ...)` and `drgn_error_from_python()`?

brenns10 replied:
In fact, I tried this and then [decided it was not great](https://github.com/osandov/drgn/pull/241#discussion_r1367556859). That basically invents a Python exception which never happened, converts it into a Drgn error, passes it down into the symbol finder layer, and then typically it gets translated back into a Python exception again (e.g. if you called `prog.symbol()`). But the translated exception is not still a ValueError, it's a generic `Exception` with the `ValueError: ...` text embedded within it.

osandov replied:
Ah, thanks for the reminder. Yeah, there's some messiness with `drgn_error_from_python()` that I still need to figure out, so this is fine.

osandov commented on libdrgn/python/program.c:
```diff
@@ -477,6 +477,73 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
+		if (err)
+			return err;
+
+		if (!drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym))) {
+			drgn_symbol_destroy(sym);
+			return &drgn_enomem;
+		}
```
`no_cleanup_ptr()` sets its argument to `NULL`, so this still leaks. We can do:

```suggestion
		if (!drgn_symbol_result_builder_add(builder, sym))
			return &drgn_enomem;
		sym = NULL; // Owned by the builder now.
```

brenns10 replied:
Ah nice catch!

osandov reviewed (changes requested):
A handful of mostly aesthetic things that I missed the first time around and a bug or two, but this looks great overall, thank you!

brenns10 commented on libdrgn/drgn.h:
```diff
@@ -930,6 +930,76 @@ struct drgn_error *drgn_program_find_symbols_by_address(struct drgn_program *pro
+ *
+ * Callbacks are called in reverse order that they were originally added. In
+ * case of a search for multiple symbols, then the results of all callbacks are
+ * concatenated. If the search is for a single symbol, then the first callback
+ * which finds a symbol will short-circuit the search.
```
That's an interesting idea. My initial reaction is that, if a caller cared that much about the specific symbol they got, they'd probably go to the effort of doing a search for multiple symbols by the same name so they could apply appropriate criteria?

I mean, I think the current logic for ELF makes perfect sense: we know the best option within the ELF symbol table is a global symbol, etc. But I can't imagine a context where a symbol finder can know that its symbol is more important than, or less important than, the result from another finder in a generally correct way.

Thinking about our proposed symbol finders: ELF, kallsyms, module exports or module kallsyms, BPF... I can't think of a precedence that would make sense for a name lookup. If the caller cares that much, they should probably use the specific symbol finder they care about directly (which, admittedly, we don't have an API for, but could add it).

osandov replied:
Yeah, I don't think this is necessary for the time being. If we need it later, we can add something like `drgn_symbol_result_builder_continue_search` (terrible name) that a finder could call before returning.

osandov commented on libdrgn/python/program.c:
```diff
@@ -477,6 +477,73 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
+	for (size_t i = 0; i < len; i++) {
+		PyObject *item = PySequence_Fast_GET_ITEM(obj, i);
+		if (!PyObject_TypeCheck(item, &Symbol_type)) {
+			PyErr_SetString(PyExc_ValueError,
```
Should this be `PyExc_TypeError`? Or `drgn_error_create(DRGN_ERROR_TYPE, ...)`?

brenns10 replied:
Yes, by my very own logic!

brenns10 replied:
```
>>> def finder(name, addr, one):
...     if name == "pop":
...         return ["goes the weasel!"]
...     return []
...
>>> prog.add_symbol_finder(finder)
>>> prog.symbol("pop")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
Exception: ValueError: symbol find callback elements must be type Symbol
```

brenns10 replied:
And after:
```
>>> prog.symbol("pop")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
TypeError: symbol find callback elements must be type Symbol
```

brenns10 replied:
And since that was really clunky wording:
```
>>> prog.symbol("pop")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
TypeError: symbol finder results must be of type Symbol
```

osandov commented on libdrgn/drgn.h:
```diff
@@ -930,6 +930,73 @@ struct drgn_error *drgn_program_find_symbols_by_address(struct drgn_program *pro
+ * The callback should perform a symbol lookup based on the flags given in @a
+ * flags. When multiple flags are provided, the effect should be treated as a
+ * logical AND. Symbol results should be added to the result builder @a builder,
+ * via @ref drgn_symbol_result_builder_add(). When @ref DRGN_FIND_SYM_ONE is
```
`s/DRGN_FIND_SYM_ONE/DRGN_FIND_SYMBOL_ONE/`

osandov commented on libdrgn/drgn.h:
```diff
@@ -930,6 +930,73 @@ struct drgn_error *drgn_program_find_symbols_by_address(struct drgn_program *pro
+ * Add or set the return value for a symbol search
+ *
+ * Symbol finders should call this with each symbol search result. If the symbol
+ * search was @ref DRGN_FIND_SYM_ONE, then only the most recent symbol added to
```
`s/DRGN_FIND_SYM_ONE/DRGN_FIND_SYMBOL_ONE/`

brenns10 replied:
You know, maybe this should be my push to use a better find-replace option than what I normally do. I use clangd's rename option all the time for renaming struct fields, etc. It's great because it is type-aware, so it can avoid doing renames on other colliding entities. But the downside is that it doesn't bother with comments.

osandov reviewed (commented):
Just a couple of comments that didn't get updated, and the commit message for "libdrgn: Use Symbol Finder API in find_symbol_by_address_internal()" is also stale: "the lookup API does allow us to keep the already-determined module" is no longer true.

Other than those tiny complaints, I'm happy with this! There's a bunch of stuff currently in `main` that I want to get in a release, but I'd like to give this new API a chance to sit for a bit before putting it in a release. So, my plan is to cut a new release early next week and then merge this immediately afterwards.

brenns10 commented:
Thanks for the last catches, clearly I need a better find/replace workflow 😅 
I updated the commit message as you mentioned, and also went ahead and reset the entire branch's author date, so that it doesn't have a crazy mixture of author dates in the log (I hate that!)

The release plan makes total sense, I wouldn't plop a this into a brand new release immediately if I were you. Looking forward to 0.0.26!

Thanks again for the detailed review, it's a much better product as a result!
